accepted/2023/NuGetAudit-supress-advisory.md

```diff
@@ -0,0 +1,230 @@
+
+<!-- Why should we not do this? -->
+- Priority
+
+The linked issue has few up-votes, so we don't yet have evidence that it's a high impact feature.
+
+- Alternatives
+
+Customers can already achieve a similar outcome, but it will require a lot more effort on their end.
+Firstly, they wouldn't use NuGetAudit, at least on CI.
+Instead, they would use `dotnet list package --format json --vulnerable` (possibly also with `--include-transitive`) as a separate step in their pipeline.
+Finally, they would need to have a script that parses the output, remove vulnerabilities they wish to suppress, then return a non-zero exit code if the result still contains any other vulnerabilities.
+
+This effort would need to be duplicated by every customer who wishes to implement such behavior.
+Even if someone in the community (or even the NuGet team ourselves) provides a sample, it requires developers to find the location where the sample is documented, and then to adapt it to their needs.
+Alternatively, if this feature is implemented, then it's much easier for customers to adopt, as it's a fairly small configuration change.
+
+## Rationale and alternatives
+
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+
+### MSBuild vs nuget.config
+
+Some NuGet features are only available via *nuget.config*, and not MSBuild items or properties.
+However, we had [one customer provide feedback that they'd like the suppression to be per-project](https://github.com/NuGet/Home/issues/11926#issuecomment-1683384471), rather than per repo.
+This allows a project already using a package with a known vulnerability to prevent CI pipeline errors, while preventing other projects in the solution from starting to use it.
+This seems like an important use-case that should be enabled.
+
+### MSBuild item vs property
+
+The current proposal is to use an MSBuild item. NuGet already uses MSBuild properties for various settings, such as additional package sources, and we can learn from experiences using it.
+
+Firstly, adding a per-project item into an MSBuild property has worse syntax (at least in my opinion): `<NuGetAuditSuppress>$(NuGetAuditSuppress);https://sample.test/CVE</NuGetAuditSuppress>`.
+
+Secondly, when there are a long list of values, it similarly ends up with a poor editing experience. Either a very long line, or an XML element with new lines as part of the value:
+
+```xml
+<NuGetAuditSuppress>
+  https://example.com/CVE1;
+  https://example.com/CVE2;
+<NuGetAuditSuppress>
+```
+
+While creating a new MSBuild item for each advisory URL uses more characters I feel it "fits" better with XML & MSBuild generally.
+
+Finally, MSBuild properties don't allow additional metadata.
+By using an item, we can add metadata, such as `Packages`: `<NuGetAuditSuppress Include="https://example.com/CVE1" Packages="ExamplePackage1;ExamplePackage2" />`.
```

Comment:
I'm not sure this is something that's desirable. It introduces a syntax not available for any other item. 

If we want suppression on a per package level, maybe a package item is more appropriate? I understand the transitive dependency limitations, but this is arguably something we're not going to add immediately anyways.

Reply:
We're not going to be adding per-package suppressions right now, so I've moved this to the 'Future Possibilities' section.